File: leetcode/c++/42_trapping_rain_water.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        int l {0};
        int r = height.size() - 1;
        int lMax {0}, rMax {0};
        int res {0};
        while (l < r) {
            if (height[l] < height[r]) {
                if (height[l] < lMax)
                    res += lMax - height[l];
                else
                    lMax = height[l];
                ++l;
            }
            else {
                if (height[r] < rMax)
                    res += rMax - height[r];
                else
                    rMax = height[r];
                --r;
            }
        }
        return res;
    }
};

int main() {
    return 0;
}

File: leetcode/rust/src/42_trapping_rain_water.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut left = vec![0; height.len()];
        let mut right = vec![0; height.len()];
        for i in 1..height.len() {
            left[i] = left[i - 1].max(height[i - 1]);
        }

        for i in (0..height.len() - 1).rev() {
            right[i] = right[i + 1].max(height[i + 1]);
        }

        let mut res = 0;
        for i in 0..height.len() {
            let min = left[i].min(right[i]);
            if min > height[i] {
                res += min - height[i];
            }
        }
        res
    }
}
